File: vue/src/lib/tenant-settings.ts
/**
 * tenant-settings.ts - 租户设置业务逻辑层
 *
 * 职责：封装租户设置（存储/AI）的状态管理与 API 操作
 * 对外暴露：useTenantSettings() composable 函数
 */
import { ref, reactive } from 'vue'
import { http } from '@/utils/http'
import type {
  TenantSettings,
  UpdateStorageReq,
  StorageUsage,
  AISettings,
  AccessSettings,
} from '@/utils/types'
import { toast } from 'vue-sonner'

const defaultStorageForm = (): UpdateStorageReq => ({
  enabled: false,
  provider: 'custom',
  endpoint: '',
  region: '',
  bucket: '',
  access_key_id: '',
  secret_access_key: '',
  custom_domain: '',
  default_target: 'local',
  force_path_style: false,
})

export function useTenantSettings() {
  const settings = ref<TenantSettings | null>(null)
  const loading = ref(false)
  const saving = ref(false)
  const testing = ref(false)

  // 存储设置表单
  const storageForm = reactive<UpdateStorageReq>(defaultStorageForm())
  const usage = ref<StorageUsage>({ used_bytes: 0, limit_bytes: 104857600, used_mb: 0, limit_mb: 100, percent: 0 })

  // AI 设置表单
  const aiForm = reactive<AISettings>({
    enabled: false,
    chat: { base_url: '', model_id: '' },
    embedding: { base_url: '', model_id: '' },
    reranker: { base_url: '', model_id: '' },
  })

  // 访问控制设置表单
  const accessForm = reactive<AccessSettings>({
    maintenance_mode: false,
    gallery_login_required: false,
  })

  async function loadSettings() {
    loading.value = true
    const res = await http.get<TenantSettings>('/tenant/settings')
    loading.value = false
    if (res.code === 0 && res.data) {
      settings.value = res.data
      // 同步到表单（不含 secret_access_key，后端不回传）
      if (res.data.storage) {
        const s = res.data.storage
        Object.assign(storageForm, {
          enabled: s.enabled,
          provider: s.provider || 'custom',
          endpoint: s.endpoint || '',
          region: s.region || '',
          bucket: s.bucket || '',
          access_key_id: s.access_key_id || '',
          secret_access_key: '',
          custom_domain: s.custom_domain || '',
          default_target: s.default_target || 'local',
          force_path_style: s.force_path_style || false,
        })
      }
      if (res.data.ai) {
        const a = res.data.ai
        Object.assign(aiForm, {
          enabled: a.enabled,
          chat: a.chat ? { ...a.chat } : { base_url: '', model_id: '' },
          embedding: a.embedding ? { ...a.embedding } : { base_url: '', model_id: '' },
          reranker: a.reranker ? { ...a.reranker } : { base_url: '', model_id: '' },
        })
      }
      if (res.data.access) {
        Object.assign(accessForm, res.data.access)
      }
    }
  }

  async function loadUsage() {
    const res = await http.get<StorageUsage>('/tenant/settings/storage/usage')
    if (res.code === 0 && res.data) {
      usage.value = res.data
    }
  }

  async function saveStorage() {
    saving.value = true
    // 若 secret_access_key 为空则不传（避免覆盖已有密钥）
    const payload: Partial<UpdateStorageReq> = { ...storageForm }
    if (!payload.secret_access_key) delete payload.secret_access_key
    const res = await http.patch('/tenant/settings/storage', payload)
    saving.value = false
    if (res.code === 0) {
      toast.success('存储设置已保存')
      loadSettings()
    } else {
      toast.error(res.message)
    }
  }

  async function testS3Connection() {
    testing.value = true
    const payload: Partial<UpdateStorageReq> = { ...storageForm }
    if (!payload.secret_access_key) delete payload.secret_access_key
    const res = await http.post('/tenant/settings/storage/test', payload)
    testing.value = false
    if (res.code === 0) {
      toast.success('云存储连接成功')
    } else {
      toast.error(`连接失败：${res.message}`)
    }
  }

  async function saveAI() {
    saving.value = true
    const res = await http.patch('/tenant/settings/ai', aiForm)
    saving.value = false
    if (res.code === 0) {
      toast.success('AI 设置已保存')
      loadSettings()
    } else {
      toast.error(res.message)
    }
  }

  async function saveAccess() {
    saving.value = true
    const res = await http.patch('/tenant/settings/access', accessForm)
    saving.value = false
    if (res.code === 0) {
      toast.success('访问控制设置已保存')
      loadSettings()
    } else {
      toast.error(res.message)
    }
  }

  function formatBytes(bytes: number): string {
    if (bytes === 0) return '0 B'
    const k = 1024
    const sizes = ['B', 'KB', 'MB', 'GB']
    const i = Math.floor(Math.log(bytes) / Math.log(k))
    return `${(bytes / Math.pow(k, i)).toFixed(1)} ${sizes[i]}`
  }

  return {
    settings,
    loading,
    saving,
    testing,
    storageForm,
    aiForm,
    usage,
    loadSettings,
    loadUsage,
    saveStorage,
    testS3Connection,
    saveAI,
    accessForm,
    saveAccess,
    formatBytes,
  }
}

File: backend/internal/model/entity/tenant_settings.go
// Package entity MongoDB 文档映射结构体
//
// 职责：定义租户功能设置集合的文档结构
// 对外接口：TenantSettings, StorageSettings, AISettings, AIEndpoint
package entity

import "time"

// TenantSettings 租户功能设置，对应 MongoDB tenant_settings 集合
// _id 直接使用 tenant_id 字符串，一租户一条记录
type TenantSettings struct {
	TenantID  string           `bson:"_id"        json:"tenant_id"`
	Storage   *StorageSettings `bson:"storage"    json:"storage"`
	AI        *AISettings      `bson:"ai"         json:"ai"`
	Access    *AccessSettings  `bson:"access"     json:"access"` // 站点级访问控制（维护模式 / 画廊登录）
	CreatedAt time.Time        `bson:"created_at" json:"created_at"`
	UpdatedAt time.Time        `bson:"updated_at" json:"updated_at"`
}

// AccessSettings 站点级访问控制配置
// 维护模式：整站仅登录可见；画廊登录：主题列表页需登录
type AccessSettings struct {
	MaintenanceMode      bool `bson:"maintenance_mode"       json:"maintenance_mode"`
	GalleryLoginRequired bool `bson:"gallery_login_required" json:"gallery_login_required"`
}

// StorageSettings 云存储 S3 配置（租户级）
type StorageSettings struct {
	Enabled         bool   `bson:"enabled"           json:"enabled"`
	Provider        string `bson:"provider"          json:"provider"` // aws / minio / aliyun / cloudflare / custom
	Endpoint        string `bson:"endpoint"          json:"endpoint"`
	Region          string `bson:"region"            json:"region"`
	Bucket          string `bson:"bucket"            json:"bucket"`
	AccessKeyID     string `bson:"access_key_id"     json:"access_key_id"`
	SecretAccessKey string `bson:"secret_access_key" json:"-"`                // 敏感字段，永不回传前端
	CustomDomain    string `bson:"custom_domain"     json:"custom_domain"`    // 可选 CDN/自定义域名
	DefaultTarget   string `bson:"default_target"    json:"default_target"`   // "local" | "cloud"
	ForcePathStyle  bool   `bson:"force_path_style"  json:"force_path_style"` // MinIO 等需要 path style
}

// AISettings AI 服务配置（租户级），预留用于后续 agent 扩展
type AISettings struct {
	Enabled   bool        `bson:"enabled"   json:"enabled"`
	Chat      *AIEndpoint `bson:"chat"      json:"chat"`
	Embedding *AIEndpoint `bson:"embedding" json:"embedding"`
	Reranker  *AIEndpoint `bson:"reranker"  json:"reranker"`
}

// AIEndpoint 单个 AI 服务端点（OpenAI 兼容接口）
type AIEndpoint struct {
	BaseURL string `bson:"base_url" json:"base_url"`
	APIKey  string `bson:"api_key"  json:"-"` // 敏感字段，永不回传前端
	ModelID string `bson:"model_id" json:"model_id"`
}
